### PR TITLE
Fix KI protection decorator

### PR DESCRIPTION
**Pull request by A5rocks**
This should fix the `trio.Lock` issue in https://github.com/python-trio/trio/issues/2775.

### CODE REVIEW

**Comment by apollo13** [comment]
Can confirm that it fixes my `Lock` issue.

**Review by TeamSpen210 — APPROVED** [review]
Makes sense as a solution, I had a little difficulty too when doing this originally. Actually though, it'd be better to use a typevar with `bound=Callable`, instead of ParamSpec - in that case, the checker can preserve the entire function type, including things like overloads.

File: trio/_core/_ki.py [hunk 1]
```diff
@@ -183,14 +183,18 @@ def wrapper(*args: ArgsT.args, **kwargs: ArgsT.kwargs) -> RetT:
     return decorator
 
 
-enable_ki_protection: Callable[
-    [Callable[ArgsT, RetT]], Callable[ArgsT, RetT]
-] = _ki_protection_decorator(True)
+# pyright work around: https://github.com/microsoft/pyright/issues/5866
```
Comment (jakkdl) [inline trio/_core/_ki.py:186]: [nit] you probably want "workaround" or "work-around"

**Review by jakkdl — APPROVED** [review]
though if teamspen's suggestion of a typevar works that sounds like it would be even better

**Comment by A5rocks** [comment]
> Makes sense as a solution, I had a little difficulty too when doing this originally. Actually though, it'd be better to use a typevar with `bound=Callable`, instead of ParamSpec - in that case, the checker can preserve the entire function type, including things like overloads.

I'm not entirely sure this is possible due to use using `*args: P.args`, etc. inside. It might be possible though? And that might be better external-facing behavior so probably worth a `# type: ignore` if CI running atm says it doesn't work.

File: trio/_core/_ki.py [hunk 2]
```diff
@@ -6,16 +6,17 @@
...
 from collections.abc import Callable
 from functools import wraps
-from typing import TYPE_CHECKING, Final, Protocol, TypeVar
+from typing import TYPE_CHECKING, Any, Final, Protocol, TypeVar
 
 import attr
 
 from .._util import is_main_thread
 
+CallableT = TypeVar("CallableT", bound=Callable[..., Any])
```
Comment (TeamSpen210) [inline trio/_core/_ki.py]: This can just use `object`, if you want to avoid the `Any`.

**Review by CoolCat467 — APPROVED** [review]

**Pull request merged**